src/syscheck.c

```diff
@@ -0,0 +1,199 @@
+    test_time_us = 5 * 1000000 / system_hz;
+    while (1) {
+        unsigned long long d;
+        if (clock_gettime(CLOCK_MONOTONIC, &ts) < 0)
+            return 0;
+        d = (ts.tv_sec * 1000000 + ts.tv_nsec / 1000) - start_us;
+        if (d >= test_time_us) break;
+    }
+    if (getrusage(RUSAGE_SELF, &ru_end) != 0)
+        return 0;
+
+    long long stime_us = (ru_end.ru_stime.tv_sec * 1000000 + ru_end.ru_stime.tv_usec) - (ru_start.ru_stime.tv_sec * 1000000 + ru_start.ru_stime.tv_usec);
+    long long utime_us = (ru_end.ru_utime.tv_sec * 1000000 + ru_end.ru_utime.tv_usec) - (ru_start.ru_utime.tv_sec * 1000000 + ru_start.ru_utime.tv_usec);
+
+    /* If more than 10% of the process time was in system calls we probably have an inefficient clocksource, print a warning */
+    if (stime_us * 10 > stime_us + utime_us) {
+        *error_msg = clocksource_warning_msg();
+        return -1;
+    } else {
+        return 1;
+    }
+}
+
+int check_xen(sds *error_msg) {
+    sds curr = read_sysfs_line("/sys/devices/system/clocksource/clocksource0/current_clocksource");
+    int res = 1;
+    if (curr == NULL) {
+        res = 0;
+    } else if (strcmp(curr, "xen") == 0) {
+        *error_msg = clocksource_warning_msg();
+        res = -1;
+    }
+    sdsfree(curr);
+    return res;
+}
+
+int check_overcommit(sds *error_msg) {
+    FILE *fp = fopen("/proc/sys/vm/overcommit_memory","r");
+    char buf[64];
+
+    if (!fp) return -1;
+    if (fgets(buf,64,fp) == NULL) {
+        fclose(fp);
+        return 0;
+    }
+    fclose(fp);
+
+    if (atoi(buf)) {
+        *error_msg = sdsnew("WARNING overcommit_memory is set to 0! Background save may fail under low memory condition. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.");
+        return -1;
+    } else {
+        return 1;
+    }
+}
+
+int check_thp_enabled(sds *error_msg) {
+    char buf[1024];
+
+    FILE *fp = fopen("/sys/kernel/mm/transparent_hugepage/enabled","r");
+    if (!fp) return 0;
+    if (fgets(buf,sizeof(buf),fp) == NULL) {
+        fclose(fp);
+        return 0;
+    }
+    fclose(fp);
+
+    if (strstr(buf,"[always]") != NULL) {
+        *error_msg = sdsnew("You have Transparent Huge Pages (THP) support enabled in your kernel. This will create latency and memory usage issues with Redis. To fix this issue run the command 'echo madvise > /sys/kernel/mm/transparent_hugepage/enabled' as root, and add it to your /etc/rc.local in order to retain the setting after a reboot. Redis must be restarted after THP is disabled (set to 'madvise' or 'never').");
+        return -1;
+    } else {
+        return 1;
+    }
+}
+#endif
+
+typedef struct {
+    const char *name;
+    int (*check_fn)(sds*);
+} check;
+
+check checks[] = {
+#ifdef __linux__
+    {.name = "clocksource", .check_fn = check_clocksource},
+    {.name = "xen", .check_fn = check_xen},
```

Comment:
do we need both `xen` and `clocksource` here?
maybe just `clocksource`?

p.s. i would rename one to `xen-clocksource` and the other to `slow-clock_gettime` or `slow-clock`

Reply:
Now that each test prints a different warning message, having both of them provides the user with more details. I'm for keeping both.